url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockMatrixOps.java

```diff
@@ -0,0 +1,674 @@
+    }
+
+    public static boolean isEquals( BlockMatrix64F A , BlockMatrix64F B )
+    {
+        if( A.blockLength != B.blockLength )
+            return false;
+
+        return MatrixFeatures.isEquals(A,B);
+    }
+
+    public static boolean isEquals( BlockMatrix64F A , BlockMatrix64F B , double tol )
+    {
+        if( A.blockLength != B.blockLength )
+            return false;
+
+        return MatrixFeatures.isEquals(A,B,tol);
+    }
+
+    /**
+     * Sets either the upper or low triangle of a matrix to zero
+     */
+    public static void zeroTriangle( boolean upper , BlockMatrix64F A )
+    {
+        int blockLength = A.blockLength;
+
+        if( upper ) {
+            for( int i = 0; i < A.numRows; i += blockLength ) {
+                int h = Math.min(blockLength,A.numRows-i);
+
+                for( int j = i; j < A.numCols; j += blockLength ) {
+                    int w = Math.min(blockLength,A.numCols-j);
+
+                    int index = i*A.numCols + h*j;
+
+                    if( j == i ) {
+                        for( int k = 0; k < h; k++ ) {
+                            for( int l = k+1; l < w; l++ ) {
+                                A.data[index + w*k+l ] = 0;
+                            }
+                        }
+                    } else {
+                        for( int k = 0; k < h; k++ ) {
+                            for( int l = 0; l < w; l++ ) {
+                                A.data[index + w*k+l ] = 0;
+                            }
+                        }
+                    }
+                }
+            }
+        } else {
+            for( int i = 0; i < A.numRows; i += blockLength ) {
+                int h = Math.min(blockLength,A.numRows-i);
+
+                for( int j = 0; j <= i; j += blockLength ) {
+                    int w = Math.min(blockLength,A.numCols-j);
+
+                    int index = i*A.numCols + h*j;
+
+                    if( j == i ) {
+                        for( int k = 0; k < h; k++ ) {
+                            int z = Math.min(k,w);
+                            for( int l = 0; l < z; l++ ) {
+                                A.data[index + w*k+l ] = 0;
+                            }
+                        }
+                    } else {
+                        for( int k = 0; k < h; k++ ) {
+                            for( int l = 0; l < w; l++ ) {
+                                A.data[index + w*k+l ] = 0;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Copies either the upper or lower triangular portion of src into dst.  Dst can be smaller
+     * than src.
+     *
+     * @param upper If the upper or lower triangle is copied.
+     * @param src The source matrix. Not modified.
+     * @param dst The destination matrix. Modified.
+     */
+    public static void copyTriangle( boolean upper , BlockMatrix64F src , BlockMatrix64F dst )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.